scripts/patch_websockets.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Patches uWebsockets Socket.h to fix a crash sending large
+payloads with ssl encoding.
+"""
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string("file", None, "Sockets.h file to patch.")
+
+# The line of Socket.h that we intend to start overwriting, and the 
+# number of lines to overwite.
+REPLACE_BLOCK_START_LINE = 308
+REPLACE_BLOCK_OVERWRITE_LENGTH = 17
+
+# The code which fixes the crash which will replace the existing
+# error handling implementation with one that retries the writes
+# when the SSL encoding buffer is full.
+REPLACE_BLOCK_CODE_LINES = [
+  "              bool continue_loop = true;\n",
+  "              do {\n",
+  "                sent = SSL_write(ssl, message->data, message->length);\n",
+  "                static int bytes_written = 0;\n",
+  "                bytes_written += sent;\n",
+  "                if (sent == (ssize_t) message->length) {\n",
+  "                  wasTransferred = false;\n",
+  "                  return true;\n",
+  "                }\n",
+  "                else if (sent < 0) {\n",
+  "                  switch (SSL_get_error(ssl, sent)) {\n",
+  "                  case SSL_ERROR_WANT_READ:\n",
+  "                    continue_loop = false;\n",
+  "                    break;\n",
+  "                  case SSL_ERROR_WANT_WRITE:\n",
+  "                    break;\n",
+  "                  default:\n",
+  "                    return false;\n",
+  "                }\n",
+  "              }\n",
+  "            }  while (continue_loop);\n"
+]
+
+def main(argv):
+  """Patches uWebSockets' Socket.h file to fix a crash when attempting
+  large writes.
+  
+  This code simply replaces one block of code (lines 308-325) with our
+  custom code defined in REPLACE_BLOCK_CODE_LINES above.
+  """
+
+  with open(FLAGS.file,'r') as sockets_file:
+    lines = sockets_file.readlines()
+    sockets_file.close()
+
+  ignore_lines_count = 0
+  with open(FLAGS.file,'w') as sockets_file:
+    index = 0
+    for line in lines:
+      if(ignore_lines_count != 0):
+        ignore_lines_count -= 1
+        index+=1
+        continue
+
+      if( index == REPLACE_BLOCK_START_LINE ):
```

Comment:
It might be worth doing a simple check for the file being patched. If versions of websockets change, this will break. Couple of options could be to not use line numbers for block replacements or just verify the block you are targetting has the expected values.

Reply:
Yes, I think this script should throw exception if the Socket.h is changed.
And in cmake file, the build should fail when this python script throw exception.

Reply:
To simplify the check, I think you can basically check if the commit in cmake/external/uWebSockets.cmake is 4d94401b9c98346f9afd838556fdc7dce30561eb

Every time when we bump up uWebSockets version, we need to check if this patch is still valid.

Reply:
I'll definitely add the check.   Smart move.  

Just a note: there are few verisons to bump to.  The library was rewritten a few versions after the one wer're using.  I think we should have a longer horizon task on our backlog to replace this uWebSockets implementation with either their latest and greatest, or a different library, since support for the version that we're using was sunset nearly 3 years ago.

Reply:
Check added.  Great idea.